pkg/sqlmodel/where_handle.go

```diff
@@ -0,0 +1,153 @@
+// Copyright 2022 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sqlmodel
+
+import (
+	"github.com/pingcap/tidb/parser/model"
+	"github.com/pingcap/tidb/parser/mysql"
+	"github.com/pingcap/tidb/types"
+)
+
+// WhereHandle is used to generate a WHERE clause in SQL.
+type WhereHandle struct {
+	UniqueNotNullIdx *model.IndexInfo
+	// If the index and columns have no NOT NULL constraint, but all data is NOT
+	// NULL, we can still use it.
+	// every index that is UNIQUE should be added to UniqueIdxs, even for
+	// PK and NOT NULL.
+	UniqueIdxs []*model.IndexInfo
+}
+
+// GetWhereHandle calculates a WhereHandle by source/target TableInfo's indices,
+// columns and state. Other component can cache the result.
```

Comment:
```suggestion
// columns and state.
```

Reply:
Just as above comment, for each row change it will calcucalte WhereHandle once. So caller may want to maintain a cache